File: Data Structures & Algorithms/word-break/submission-8.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        dp = {}
        N = len(s)
        def dfs(i):
            if i == N:
                return True
            if i in dp:
                return dp[i]
            for w in wordDict:
                if i+len(w)<=N and s[i:i+len(w)] == w:
                    dp[i] = dfs(i+len(w))
                    if dp[i]:
                        return dp[i]
            dp[i] = False
            return dp[i]
        return dfs(0)


        

File: Data Structures & Algorithms/combination-target-sum-ii/submission-8.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> comb = new ArrayList<>();

        dfs(0,0,target,candidates,comb,res);
        return res;
        
    }

    public void dfs(int i, int s, int target, int[] candidates, 
    List<Integer> comb,List<List<Integer>> res) {
        if(s==target){
            res.add(new ArrayList<>(comb));
            return;
        }
        if(i==candidates.length || s>target){
            return;
        }
        comb.add(candidates[i]);
        dfs(i+1,s+candidates[i],target,candidates,comb,res);
        comb.remove(comb.size()-1);
        while(i+1<candidates.length && candidates[i] == candidates[i+1]){
            i++;
        }
        dfs(i+1,s,target,candidates,comb,res);
    }
}
